// Copyright (c) 2022 AccelByte Inc. All Rights Reserved.
// This is licensed software from AccelByte Inc, for limitations
// and restrictions contact your company contract manager.

syntax = "proto3";

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";

package matchfunctiongrpc;

// GetStatCodes
message GetStatCodesRequest {
  Rules rules = 1;
}
message StatCodesResponse {
  repeated string codes = 1;
}

service MatchFunction {
  rpc GetStatCodes(GetStatCodesRequest) returns (StatCodesResponse);
  rpc ValidateTicket(ValidateTicketRequest) returns (ValidateTicketResponse);
  rpc MakeMatches(stream MakeMatchesRequest) returns (stream MatchResponse);
}

// ValidateTicket
message ValidateTicketRequest {
  Ticket ticket = 1;
  Rules rules = 2;
}
message ValidateTicketResponse {
  bool valid = 1;
}

// Make Matches
message MakeMatchesRequest {
  message MakeMatchesParameters{
    Scope scope = 1;
    Rules rules = 2;
  }
  oneof request_type {
    MakeMatchesParameters parameters = 1;
    Ticket ticket = 2;
  }
}
message MatchResponse {
  Match match = 1;
}

message Match {
  message Team {
    repeated string user_ids = 1;
  }
  repeated Ticket tickets = 1;
  repeated Team teams = 2;
  repeated string region_preferences = 3;
  google.protobuf.Struct match_attributes = 4;
}

message Rules {
  string json = 1;
}

message Scope {
  string ab_trace_id = 1;
}

message Ticket {
  message PlayerData {
    string player_id = 1;
    google.protobuf.Struct attributes = 2;
  }
  string ticket_id = 1;
  string match_pool = 2;
  google.protobuf.Timestamp CreatedAt = 3;
  repeated PlayerData players = 4;
  google.protobuf.Struct ticket_attributes = 5;
  map<string, int64> latencies = 6;
}
